### ISSUE TITLE
Incorrect result returned by `UNION`

### ISSUE THREAD

**Issue by YuanchengJiang** [crate/crate#14807]
### CrateDB version

5.4.3

### CrateDB setup information

Table Info:
```
create table union_test (c_0 timestamp, c_1 int, c_2 float, c_3 varchar(16));
insert into union_test values (946702800000, 8, 3.0, 'Goo');
insert into union_test values (946688400000, 9, 7.0, 'Boo');
insert into union_test values (946692000000, 2, 4.0, 'Foo');
insert into union_test values (946695600000, 6, 8.0, 'Boo');
insert into union_test values (946699200000, 7, 6.0, 'Coo');
```

### Problem description

The result returned by `UNION` is not correct.

Q1=`SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 CROSS JOIN union_test as T2 INNER JOIN union_test as T3 ON T2.c_0=T3.c_0`

Result:
```
cr> SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 CROSS JOIN union_t
    est as T2 INNER JOIN union_test as T3 ON T2.c_0=T3.c_0;
+------------------------------------+------------------------------------+------------------------------------+
| collection_count(collect_set(c_0)) | collection_count(collect_set(c_1)) | collection_count(collect_set(c_1)) |
+------------------------------------+------------------------------------+------------------------------------+
|                                  5 |                                  5 |                                  5 |
+------------------------------------+------------------------------------+------------------------------------+
SELECT 1 row in set (0.029 sec)
```

Q2=`SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 JOIN union_test as T2 ON T1.c_1=T2.c_1 INNER JOIN union_test as T3 ON T2.c_0<T3.c_0 group by T1.c_1`

Result:
```
cr> SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 JOIN union_test as
     T2 ON T1.c_1=T2.c_1 INNER JOIN union_test as T3 ON T2.c_0<T3.c_0 group by T1.c_1;
+------------------------------------+------------------------------------+------------------------------------+
| collection_count(collect_set(c_0)) | collection_count(collect_set(c_1)) | collection_count(collect_set(c_1)) |
+------------------------------------+------------------------------------+------------------------------------+
|                                  1 |                                  1 |                                  4 |
|                                  1 |                                  1 |                                  3 |
|                                  1 |                                  1 |                                  2 |
|                                  1 |                                  1 |                                  1 |
+------------------------------------+------------------------------------+------------------------------------+
SELECT 4 rows in set (0.052 sec)
```

Q=`Q1 UNION ALL Q2`

Result:
```
cr> SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 CROSS JOIN union_t
    est as T2 INNER JOIN union_test as T3 ON T2.c_0=T3.c_0 UNION ALL SELECT count (distinct T1.c_0), count (distinct T2.c_1),
     count (distinct T3.c_1) FROM union_test as T1 JOIN union_test as T2 ON T1.c_1=T2.c_1 INNER JOIN union_test as T3 ON T2.c
    _0<T3.c_0 group by T1.c_1;
+------------------------------------+------------------------------------+------------------------------------+
| collection_count(collect_set(c_0)) | collection_count(collect_set(c_1)) | collection_count(collect_set(c_1)) |
+------------------------------------+------------------------------------+------------------------------------+
|                                  5 |                                  5 |                                  5 |
|                                  1 |                                  1 |                                  1 |
|                                  1 |                                  1 |                                  1 |
|                                  1 |                                  1 |                                  1 |
|                                  1 |                                  1 |                                  1 |
+------------------------------------+------------------------------------+------------------------------------+
SELECT 5 rows in set (0.051 sec
```

### Steps to Reproduce

Run
```
SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 CROSS JOIN union_test as T2 INNER JOIN union_test as T3 ON T2.c_0=T3.c_0 UNION ALL SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 JOIN union_test as T2 ON T1.c_1=T2.c_1 INNER JOIN union_test as T3 ON T2.c_0<T3.c_0 group by T1.c_1;
```

### Actual Result

```
cr> SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 CROSS JOIN union_t
    est as T2 INNER JOIN union_test as T3 ON T2.c_0=T3.c_0 UNION ALL SELECT count (distinct T1.c_0), count (distinct T2.c_1),
     count (distinct T3.c_1) FROM union_test as T1 JOIN union_test as T2 ON T1.c_1=T2.c_1 INNER JOIN union_test as T3 ON T2.c
    _0<T3.c_0 group by T1.c_1;
+------------------------------------+------------------------------------+------------------------------------+
| collection_count(collect_set(c_0)) | collection_count(collect_set(c_1)) | collection_count(collect_set(c_1)) |
+------------------------------------+------------------------------------+------------------------------------+
|                                  5 |                                  5 |                                  5 |
|                                  1 |                                  1 |                                  1 |
|                                  1 |                                  1 |                                  1 |
|                                  1 |                                  1 |                                  1 |
|                                  1 |                                  1 |                                  1 |
+------------------------------------+------------------------------------+------------------------------------+
SELECT 5 rows in set (0.051 sec
```

### Expected Result

```
count | count | count 
-------+-------+-------
     1 |     1 |     3
     1 |     1 |     2
     1 |     1 |     1
     1 |     1 |     4
     5 |     5 |     5
(5 rows)

```

**Comment by jeeminso** [comment]
Thank you a lot for reporting @YuanchengJiang . It looks like repeat use of the same column names is triggering this bug. For this scenario, providing column aliases seems to avoid it:

```
cr> SELECT count (distinct T1.c_0) as x, count (distinct T2.c_1) as y, count (distinct T3.c_1) as z FROM union_test as T1 CROSS JOIN union_test as
     T2 INNER JOIN union_test as T3 ON T2.c_0=T3.c_0 
    union all 
    SELECT count (distinct T1.c_0), count (distinct T2.c_1), count (distinct T3.c_1) FROM union_test as T1 JOIN union_test as T2 ON T1.c_1=T2.c_1 
    INNER JOIN union_test as T3 ON T2.c_0<T3.c_0 group by T1.c_1;                                                                                 
+---+---+---+
| x | y | z |
+---+---+---+
| 5 | 5 | 5 |
| 1 | 1 | 4 |
| 1 | 1 | 3 |
| 1 | 1 | 2 |
| 1 | 1 | 1 |
+---+---+---+
SELECT 5 rows in set (0.006 sec)
```

**Comment by seut** [comment]
@YuanchengJiang
The issue has been fixed by #14816 and will be released with the next `5.4.5` hotfix release.
Thank you for reporting.
